### ISSUE TITLE
Issue with applying optimizations to mlir code generated by onnx-mlir from onnx file

### ISSUE THREAD

**Issue by alecerio** [onnx/onnx-mlir#2409]
# Issue description

I am encountering an issue while trying to apply optimizations to the MLIR code generated by onnx-mlir. Specifically, when attempting to run the optimization passes with the mlir-opt command, I am facing the following error:

```bash
...
DB93D7F7B54BDFD76B43C613D853CA1A8173DDD540CBC77968A3C08F24BBD204459BD1566F43C7AEB83BDDE9D7B3CE367B63CA948063D0A4F133DF0FB293D717CE6BC3F7422BD63174F3D0813AEBCA14C303B9A9813BD49A8D43CC86B85BDE07469BD11148EBDB6BFC9BC47F3A13CBA2652BDBFF526BDC8EA813DD9A0E23C29955DBD7A1874BA0D72973CD7719D3CC29C85BD96EF313D62D509BDE2B8ABBD3E07BCBD7D601ABD47570EBDF1BA0F3D1835293DE2C587BD"> : tensor<128x196xf32>} : () -> memref<128x196xf32>
```

# how to reproduce the error

To reproduce the error, you can use the following steps:

1. Install onnx-mlir from the docker file provided in the repository.
2. Emit the mlir code:
```bash
onnx-mlir --EmitMLIR mnist.onnx
```
3. Apply optimizations using mlir-opt:
```bash
mlir-opt -cse -o mnist_opt.mlir mnist.onnx.mlir
```
4. Here I get the error previously reported.

# additional notes

- The ONNX model used for testing is the mnist.onnx file from the onnx-mlir repository (see: https://github.com/onnx/onnx-mlir/blob/main/docs/mnist_example/README.md).
- I have also tried using the --allow-unregistered-dialect option to proceed with optimizations, but this leads to issues during the conversion to LLVM IR.
- I have check the dialects used in the generated mlir code, such as affine, memref, arith, and math. They are part of the core MLIR distribution and should be registered by default. Therefore, it is not clear if the problem is due to unregistered dialects.

# system info

OS: Ubuntu 22.04

Thank you for you attention.

**Comment by AlexandreEichenberger** [comment]
This path is not used much, and there are a couple of custom (KrnlDialect) ops that escaped into the mlir pass. 

For this example, here are the offending ops:

``` mlir
    %0 = "krnl.global"() {name = "constant_1", shape = [128, 196], value = dense_resource<__elided__> : tensor<128x196xf32>} : () -> memref<128x196xf32>
    %1 = "krnl.global"() {name = "constant_2", shape = [128], value = dense_resource<__elided__> : tensor<128xf32>} : () -> memref<128xf32>
    %2 = "krnl.global"() {name = "constant_3", shape = [10, 128], value = dense_resource<__elided__> : tensor<10x128xf32>} : () -> memref<10x128xf32>
    %3 = "krnl.global"() {name = "constant_4", shape = [10], value = dense<[0.0360819325, 0.0303174332, -0.014957685, 0.0604293533, 0.0917338133, 0.0978015586, -0.0705472528, 0.121654645, -0.039008636, -0.0469057448]> : tensor<10xf32>} : () -> memref<10xf32>
  "krnl.entry_point"() {func = @main_graph, numInputs = 1 : i32, numOutputs = 1 : i32, signature = "[    { \22type\22 : \22f32\22 , \22dims\22 : [1 , 1 , 28 , 28] , \22name\22 : \22image\22 }\0A\0A]\00@[   { \22type\22 :
```

@sorenlassen , @tungld : you are more familiar than me with the handling of constants; would there be a way to convert them to non-krnl based ops simply? 

@alecerio : if this path is important for you, would you be able to help if there is a need to implement new functionality for this? There may already be support (aka we also export to Tosa) but it may need some massaging. You should look at the passes that transform ONNX all the way to MLIR (as activated by this flag) and see if there are other Krnl OPS that are deemed legal by  the passes associated with this flag.

**Comment by alecerio** [comment]
Thanks for the answer @AlexandreEichenberger 

My goal is to use an onnx file (exported from a pytorch model) and convert it to MLIR to apply optimizations, then convert it to llvm ir and finally generate assembly code for a target architecture. I would like to point out that I am new to mlir and maybe the path I propose here is not the most appropriate.
With the option --EmitLLVMIR, I could convert directly to llvm dialect and consequently to llvm ir with mlir-translate. Nevertheless, this approach prevents to apply optimizations as I convert directly to the llvm dialect.
I would be glad to contribute, if this pipeline makes sense to convert onnx files to assembly code.
If instead there is a better way to proceed, just let me know.
Thanks.

**Comment by AlexandreEichenberger** [comment]
Sure, let's see if @sorenlassen or @tungld have a quick solution for converting the `krnl.global` to an mlir equivalent.

**Comment by tungld** [comment]
Even if we converting `krnl.global` to an mlir equivalent, we still have other krnl ops like `krnl.call`, and some `krnl` math ops that go to llvm ir directly. 

From the log here, it's still difficult to know whether the error is because of `krnl.global` or not. @alecerio could you print out more information related to the error? and have you tried to use `onnx-mlir-opt` instead of `mlir-opt`?

**Comment by alecerio** [comment]
- **Using onnx-mlir-opt command**

I was not aware of _onnx-mlir-opt_, now it seems to work smoothly, thanks.

After emitting mlir code with onnx-mlir --EmitMLIR, I applied some optimizations (e.g. --cse) with onnx-mlir-opt and, finally, to lower to llvm dialect I used:

```bash
onnx-mlir-opt --convert-krnl-to-llvm -o mnist_opt_llvm.mlir mnist_opt.mlir
```

With _translate-mlir_ command I could convert from llvm dialect to llvm ir without problems.

- **error log**

I just omitted the rest of the error log as it is a very long sequence of hexadecimal digits and the only human readable part was the "tensor<128x196xf32>} : () -> memref<128x196xf32>" at the end of the log.
